File: modules/Scripting/modules/RmlUi/src/IDs.cpp
#include "IDs.hpp"
#include "RmlUi/Core/ID.h"

#include "Sol2HelperMacros.hpp"

using namespace Rml;

namespace {

void registerShorthandId(sol::table &lua) {
#define MAKE_PAIR(Value) _MAKE_PAIR(ShorthandId, Value)
  // clang-format off
  lua.DEFINE_ENUM(ShorthandId, {
    MAKE_PAIR(Invalid),
    MAKE_PAIR(Margin),
    MAKE_PAIR(Padding),
    MAKE_PAIR(BorderWidth),
    MAKE_PAIR(BorderColor),
    MAKE_PAIR(BorderTop),
    MAKE_PAIR(BorderRight),
    MAKE_PAIR(BorderBottom),
    MAKE_PAIR(BorderLeft),
    MAKE_PAIR(Border),
    MAKE_PAIR(BorderRadius),
    MAKE_PAIR(Overflow),
    MAKE_PAIR(Background),
    MAKE_PAIR(Font),
    MAKE_PAIR(Gap),
    MAKE_PAIR(PerspectiveOrigin),
    MAKE_PAIR(TransformOrigin),
    MAKE_PAIR(Flex),
    MAKE_PAIR(FlexFlow),

    MAKE_PAIR(NumDefinedIds),
    MAKE_PAIR(FirstCustomId),

    MAKE_PAIR(MaxNumIds),
  });
  // clang-format on
#undef MAKE_PAIR
}
void registerPropertyId(sol::table &lua) {
#define MAKE_PAIR(Value) _MAKE_PAIR(PropertyId, Value)
  // clang-format off
  lua.DEFINE_ENUM(PropertyId, {
    MAKE_PAIR(Invalid),

    MAKE_PAIR(MarginTop),
    MAKE_PAIR(MarginRight),
    MAKE_PAIR(MarginBottom),
    MAKE_PAIR(MarginLeft),
    MAKE_PAIR(PaddingTop),
    MAKE_PAIR(PaddingRight),
    MAKE_PAIR(PaddingBottom),
    MAKE_PAIR(PaddingLeft),
    MAKE_PAIR(BorderTopWidth),
    MAKE_PAIR(BorderRightWidth),
    MAKE_PAIR(BorderBottomWidth),
    MAKE_PAIR(BorderLeftWidth),
    MAKE_PAIR(BorderTopColor),
    MAKE_PAIR(BorderRightColor),
    MAKE_PAIR(BorderBottomColor),
    MAKE_PAIR(BorderLeftColor),
    MAKE_PAIR(BorderTopLeftRadius),
    MAKE_PAIR(BorderTopRightRadius),
    MAKE_PAIR(BorderBottomRightRadius),
    MAKE_PAIR(BorderBottomLeftRadius),
    MAKE_PAIR(Display),
    MAKE_PAIR(Position),
    MAKE_PAIR(Top),
    MAKE_PAIR(Right),
    MAKE_PAIR(Bottom),
    MAKE_PAIR(Left),
    MAKE_PAIR(Float),
    MAKE_PAIR(Clear),
    MAKE_PAIR(BoxSizing),
    MAKE_PAIR(ZIndex),
    MAKE_PAIR(Width),
    MAKE_PAIR(MinWidth),
    MAKE_PAIR(MaxWidth),
    MAKE_PAIR(Height),
    MAKE_PAIR(MinHeight),
    MAKE_PAIR(MaxHeight),
    MAKE_PAIR(LineHeight),
    MAKE_PAIR(VerticalAlign),
    MAKE_PAIR(OverflowX),
    MAKE_PAIR(OverflowY),
    MAKE_PAIR(Clip),
    MAKE_PAIR(Visibility),
    MAKE_PAIR(BackgroundColor),
    MAKE_PAIR(Color),
    MAKE_PAIR(CaretColor),
    MAKE_PAIR(ImageColor),
    MAKE_PAIR(FontFamily),
    MAKE_PAIR(FontStyle),
    MAKE_PAIR(FontWeight),
    MAKE_PAIR(FontSize),
    MAKE_PAIR(TextAlign),
    MAKE_PAIR(TextDecoration),
    MAKE_PAIR(TextTransform),
    MAKE_PAIR(WhiteSpace),
    MAKE_PAIR(WordBreak),
    MAKE_PAIR(RowGap),
    MAKE_PAIR(ColumnGap),
    MAKE_PAIR(Cursor),
    MAKE_PAIR(Drag),
    MAKE_PAIR(TabIndex),
    MAKE_PAIR(ScrollbarMargin),
    MAKE_PAIR(OverscrollBehavior),

    MAKE_PAIR(Perspective),
    MAKE_PAIR(PerspectiveOriginX),
    MAKE_PAIR(PerspectiveOriginY),
    MAKE_PAIR(Transform),
    MAKE_PAIR(TransformOriginX),
    MAKE_PAIR(TransformOriginY),
    MAKE_PAIR(TransformOriginZ),

    MAKE_PAIR(Transition),
    MAKE_PAIR(Animation),

    MAKE_PAIR(Opacity),
    MAKE_PAIR(PointerEvents),
    MAKE_PAIR(Focus),

    MAKE_PAIR(Decorator),
    MAKE_PAIR(FontEffect),

    MAKE_PAIR(FillImage),

    MAKE_PAIR(AlignContent),
    MAKE_PAIR(AlignItems),
    MAKE_PAIR(AlignSelf),
    MAKE_PAIR(FlexBasis),
    MAKE_PAIR(FlexDirection),
    MAKE_PAIR(FlexGrow),
    MAKE_PAIR(FlexShrink),
    MAKE_PAIR(FlexWrap),
    MAKE_PAIR(JustifyContent),

    MAKE_PAIR(NumDefinedIds),
    MAKE_PAIR(FirstCustomId),

    MAKE_PAIR(MaxNumIds),
  });
  // clang-format on
#undef MAKE_PAIR
}
void registerEventId(sol::table &lua) {
#define MAKE_PAIR(Value) _MAKE_PAIR(EventId, Value)
  // clang-format off
  lua.DEFINE_ENUM(EventId, {
    MAKE_PAIR(Invalid),

	  MAKE_PAIR(Mousedown),
	  MAKE_PAIR(Mousescroll),
	  MAKE_PAIR(Mouseover),
	  MAKE_PAIR(Mouseout),
	  MAKE_PAIR(Focus),
	  MAKE_PAIR(Blur),
	  MAKE_PAIR(Keydown),
	  MAKE_PAIR(Keyup),
	  MAKE_PAIR(Textinput),
	  MAKE_PAIR(Mouseup),
	  MAKE_PAIR(Click),
	  MAKE_PAIR(Dblclick),
	  MAKE_PAIR(Load),
	  MAKE_PAIR(Unload),
	  MAKE_PAIR(Show),
	  MAKE_PAIR(Hide),
	  MAKE_PAIR(Mousemove),
	  MAKE_PAIR(Dragmove),
	  MAKE_PAIR(Drag),
	  MAKE_PAIR(Dragstart),
	  MAKE_PAIR(Dragover),
	  MAKE_PAIR(Dragdrop),
	  MAKE_PAIR(Dragout),
	  MAKE_PAIR(Dragend),
	  MAKE_PAIR(Handledrag),
	  MAKE_PAIR(Resize),
	  MAKE_PAIR(Scroll),
	  MAKE_PAIR(Animationend),
	  MAKE_PAIR(Transitionend),

	  MAKE_PAIR(Change),
	  MAKE_PAIR(Submit),
	  MAKE_PAIR(Tabchange),
	  MAKE_PAIR(Columnadd),
	  MAKE_PAIR(Rowadd),
	  MAKE_PAIR(Rowchange),
	  MAKE_PAIR(Rowremove),
	  MAKE_PAIR(Rowupdate),

	  MAKE_PAIR(NumDefinedIds),

	  MAKE_PAIR(FirstCustomId),

	  MAKE_PAIR(MaxNumIds),
  });
  // clang-format on
#undef MAKE_PAIR
}

} // namespace

void registerIDs(sol::table &lua) {
  registerShorthandId(lua);
  registerPropertyId(lua);
  registerEventId(lua);
}

File: modules/Scripting/annotations/RmlUi/IDs.lua
---@meta

---@enum ui.ShorthandId
ui.ShorthandId = {
  Invalid = 0,

  ---The following values define the shorthand ids for the main stylesheet specification.
  ---These values must not be used in places that have their own property specification,
  ---such as decorators and font-effects.

  Margin = 1,
  Padding = 2,
  BorderWidth = 3,
  BorderColor = 4,
  BorderTop = 5,
  BorderRight = 6,
  BorderBottom = 7,
  BorderLeft = 8,
  Border = 9,
  BorderRadius = 10,
  Overflow = 11,
  Background = 12,
  Font = 13,
  Gap = 14,
  PerspectiveOrigin = 15,
  TransformOrigin = 16,
  Flex = 17,
  FlexFlow = 18,

  NumDefinedIds = 19,
  FirstCustomId = ui.ShorthandId.NumDefinedIds,

  ---The maximum number of IDs. This limits the number of possible custom IDs to MaxNumIds - FirstCustomId.
  MaxNumIds = 0xff,
}

---@enum ui.PropertyId
ui.PropertyId = {
  Invalid = 0,

  ---The following values define the property ids for the main stylesheet specification.
  ---These values must not be used in places that have their own property specification,
  ---such as decorators and font-effects.

  MarginTop = 1,
  MarginRight = 2,
  MarginBottom = 3,
  MarginLeft = 4,
  PaddingTop = 5,
  PaddingRight = 6,
  PaddingBottom = 7,
  PaddingLeft = 8,
  BorderTopWidth = 9,
  BorderRightWidth = 10,
  BorderBottomWidth = 11,
  BorderLeftWidth = 12,
  BorderTopColor = 13,
  BorderRightColor = 14,
  BorderBottomColor = 15,
  BorderLeftColor = 16,
  BorderTopLeftRadius = 17,
  BorderTopRightRadius = 18,
  BorderBottomRightRadius = 19,
  BorderBottomLeftRadius = 20,
  Display = 21,
  Position = 22,
  Top = 23,
  Right = 24,
  Bottom = 25,
  Left = 26,
  Float = 27,
  Clear = 28,
  BoxSizing = 29,
  ZIndex = 30,
  Width = 31,
  MinWidth = 32,
  MaxWidth = 33,
  Height = 34,
  MinHeight = 35,
  MaxHeight = 36,
  LineHeight = 37,
  VerticalAlign = 38,
  OverflowX = 39,
  OverflowY = 40,
  Clip = 41,
  Visibility = 42,
  BackgroundColor = 43,
  Color = 44,
  CaretColor = 45,
  ImageColor = 46,
  FontFamily = 47,
  FontStyle = 48,
  FontWeight = 49,
  FontSize = 50,
  TextAlign = 51,
  TextDecoration = 52,
  TextTransform = 53,
  WhiteSpace = 54,
  WordBreak = 55,
  RowGap = 56,
  ColumnGap = 57,
  Cursor = 58,
  Drag = 59,
  TabIndex = 60,
  ScrollbarMargin = 61,
  OverscrollBehavior = 62,

  Perspective = 63,
  PerspectiveOriginX = 64,
  PerspectiveOriginY = 65,
  Transform = 66,
  TransformOriginX = 67,
  TransformOriginY = 68,
  TransformOriginZ = 69,

  Transition = 70,
  Animation = 71,

  Opacity = 72,
  PointerEvents = 73,
  Focus = 74,

  Decorator = 75,
  FontEffect = 76,

  FillImage = 77,

  AlignContent = 78,
  AlignItems = 79,
  AlignSelf = 80,
  FlexBasis = 81,
  FlexDirection = 82,
  FlexGrow = 83,
  FlexShrink = 84,
  FlexWrap = 85,
  JustifyContent = 86,

  NumDefinedIds = 87,
  FirstCustomId = ui.PropertyId.NumDefinedIds,

  ---The maximum number of IDs. This limits the number of possible custom IDs to MaxNumIds - FirstCustomId.
  MaxNumIds = 128,
}

---@enum ui.EventId
ui.EventId = {
  Invalid = 0,

  --- Core events:

  Mousedown = 1,
  Mousescroll = 2,
  Mouseover = 3,
  Mouseout = 4,
  Focus = 5,
  Blur = 6,
  Keydown = 7,
  Keyup = 8,
  Textinput = 9,
  Mouseup = 10,
  Click = 11,
  Dblclick = 12,
  Load = 13,
  Unload = 14,
  Show = 15,
  Hide = 16,
  Mousemove = 17,
  Dragmove = 18,
  Drag = 19,
  Dragstart = 20,
  Dragover = 21,
  Dragdrop = 22,
  Dragout = 23,
  Dragend = 24,
  Handledrag = 25,
  Resize = 26,
  Scroll = 27,
  Animationend = 28,
  Transitionend = 29,

  --- Form control events:

  Change = 30,
  Submit = 31,
  Tabchange = 32,

  NumDefinedIds = 33,

  --- Custom IDs start here.
  FirstCustomId = ui.EventId.NumDefinedIds,

  --- The maximum number of IDs. This limits the number of possible custom IDs to MaxNumIds - FirstCustomId.
  MaxNumIds = 0xffff,
}
